README.md
# Description

The Chef WordPress cookbook installs and configures WordPress according to the instructions at http://codex.wordpress.org/Installing_WordPress.

Additionally, the `setup` recipe of this cookbook uses the WordPress CLI (http://wp-cli.org/) to install the WordPress core blog.
For manual setup, use the `install` recipe and go to: http://HOSTNAME:80/wp-admin/install.php

# Requirements

## Platform

* Ubuntu 12.04, 14.04
* RHEL/CentOS 5, 6
* Windows

## Cookbooks

* mysql
* mysql_chef_gem
* php
* apache2
* windows
* openssl (uses library to generate secure passwords)
* selinux (used to disable selinux for MySQL on RHEL-based systems)

# Attributes

### WordPress

* `node['wordpress']['version']` - Version of WordPress to download. Use 'latest' to download most recent version.
* `node['wordpress']['parent_dir']` - Parent directory to where WordPress will be extracted. (Windows Only)
* `node['wordpress']['dir']` - Location to place WordPress files.
* `node['wordpress']['db']['instance_name']` - Name of the MySQL instance to use with MySQL (community cookbook version 6+)
* `node['wordpress']['db']['name']` - Name of the WordPress MySQL database.
* `node['wordpress']['db']['user']` - Name of the WordPress MySQL user.
* `node['wordpress']['db']['prefix']` - Prefix of all MySQL tables created by WordPress.
* `node['wordpress']['db']['host']` - Host of the WordPress MySQL database.
* `node['wordpress']['db']['port']` - Port of the WordPress MySQL database.
* `node['wordpress']['db']['charset']` - [Character set](http://dev.mysql.com/doc/refman/5.7/en/charset-charsets.html) of the WordPress MySQL database tables. Defaults to 'utf8'.
* `node['wordpress']['db']['collate']` - [Collation](http://dev.mysql.com/doc/refman/5.7/en/charset-collation-effect.html) of the WordPress MySQL database tables.
* `node['wordpress']['db']['mysql_version']` - Version of MySQL to install (for supporting community cookbook version 6+)

* `node['wordpress']['allow_multisite']` - Enable [multisite](http://codex.wordpress.org/Create_A_Network) features (default: false).
* `node['wordpress']['wp_config_options']` - A hash of options to define in wp_config.php, output as key value pairs into a PHP constant e.g. `define( '<%= @key %>', <%= @value %> );`. Note: for values you will need to add single quotes around text but omit them for booleans and numbers. (default: {}).
* `node['wordpress']['config_perms']` - Permissions to set for a site's wp-config.php.
* `node['wordpress']['server_aliases']` - Aliases to use when setting up Virtual Host with Nginx or Apache
* `node['wordpress']['server_port']` - Port to use when setting up the Virtual Host with Nginx or Apache

* `node['wordpress']['install']['user']` - Install user used for WordPress file permissions and the PHP-FPM user (if applicable)
* `node['wordpress']['install']['group']` - Install group used for WordPress file permissions and the PHP-FPM group (if necessary)

* `node['wordpress']['parent_dir']` - Parent directory of where WordPress will be installed. This is used in the Windows installation to determine where the .zip will be downloaded to.
* `node['wordpress']['dir']` - Path where WordPress should be installed
* `node['wordpress']['url']` - URL to the zip or tarball installer of WordPress
* `node['wordpress']['server_name']` - Hostname used for setting up the Virtual Host configuration for your WordPress site

* `node['wordpress']['php_options']` - Additional PHP settings for the installation.

See `attributes/default.rb` for further details and examples.

# Usage

Add the "wordpress" recipe to your node's run list or role, or include the recipe in another cookbook.

## Vagrant

1) Install Vagrant: https://www.vagrantup.com/
2) Install the Vagrant plugin `vagrant-omnibus` via `vagrant plugin install vagrant-omnibus`

### Virtualbox

```shell
cd vagrant-virtualbox
vagrant up
```

### Amazon EC2

1) Adjust `vagrant-aws/Vagrantfile`, especially:
  * `SSH_KEY_PATH`
  * `aws.keypair_name`
  * `aws.security_groups` (make sure HTTP port is open!)

```shell
cd vagrant-aws
export AWS_ACCESS_KEY_ID=???
export AWS_SECRET_ACCESS_KEY=???
vagrant up
```

## TestKitchen

http://kitchen.ci/

```shell
kitchen verify
```

# License and Author

* Author:: Barry Steinglass (barry@opscode.com)
* Author:: Joshua Timberman (joshua@opscode.com)
* Author:: Seth Chisamore (schisamo@opscode.com)
* Author:: Lucas Hansen (lucash@opscode.com)
* Author:: Julian C. Dunn (jdunn@getchef.com)
* Contributor:: Antek S. Baranski (antek.baranski@gmail.com)
* Contributor:: Joel Scheuner (joel.scheuner.dev@gmail.com)

Copyright:: 2010-2013, Chef Software, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.


CHANGELOG.md
# CHANGELOG

v3.2.0 (2017-01-20)
-------------------

### New Feature
* Add `setup` recipe to install both: WordPress (see `install`) and the WordPress core (see `install_core`) via WordPress CLI.
* Add Vagrantfiles (https://www.vagrantup.com/) for Virtualbox (`vagrant-virtualbox`) and Amazon EC2 (`vagrant-aws`)

### Bug
* Fix reprovisioning after VM restart (observed on Ubuntu 14.04)
* Fix to enable wordpress apache config

v3.0.0 (2015-02-24)
-------------------
* Cookbook updated to support the MySQL 6.0 community cookbook
* Added selinux as a dependency to resolve MySQL installation issues on CentOS
* Adjustments made to Nginx/PHP-FPM configuration to work out of the box with CentOS
* Added attribute to allow for additional WordPress configuration options
* Updated bats tests to work when testing WordPress version 4+ and <4
* Removed testing for Ubuntu 10.04 due to fast approaching EOL and bugs in several cookbooks WordPress depends on.

v2.1.5 (2014-06-28)
-------------------
* Recipe added to support installation with nginx

v2.1.2 (2014-04-03)
-------------------
* Fixes issue where `mysql::ruby` recipe was removed in v5 of community cookbook

v2.1.0 (2014-02-27)
-------------------
[COOK-4354] s/Wordpress/WordPress/g in docs
[COOK-4356] Gets WordPress to be extracted to the proper directory
[COOK-4191] WordPress table_prefix not configured when node['wordpress']['db']['prefix'] is set
[COOK-4192] Add attribute to control WordPress multisite features
[COOK-4366] apache2 port parameter


v2.0.0 (2014-02-27)
-------------------
[COOK-4180] leverage the database cookbook


v1.3.2 (2014-01-23)
-------------------
* [COOK-4248] - use "no_managed_code" when setting up WordPress Pool on Windows
* [COOK-4170] - Wordpress tarball contains a wordpress subdirectory, causing "extract-wordpress" to execute every run and the WP URL to be http://hostname/wordpress/


v1.3.0
------
### Bug
- **[COOK-3478](https://tickets.opscode.com/browse/COOK-3478)** - Windows support for Wordpress


v1.2.0
------
### New Feature
- **[COOK-3321](https://tickets.opscode.com/browse/COOK-3321)** - Add languages recipe

### Improvement
- **[COOK-3311](https://tickets.opscode.com/browse/COOK-3311)** - Remove legacy Test Kitchen, Add Travis CI


v1.1.0
------
- Added Test Kitchen 1.0

Bug Fixes:
- [COOK-1393]: wordpress recipe should use mysql::ruby to ensure ruby extension is installed
- [COOK-2984]: wordpress cookbook has foodcritic failures

Improvements:
- [COOK-2661]: Allow downloads from other repos for wordpress install

v1.0.0:
-------
- [COOK-1127] - update defaults to latest version
- [COOK-1222] - support installing "latest" version
- [COOK-1271] - Wordpress cookbook generates new password on every chef run

v0.8.8
------
- [COOK-826] -  recipe doesn't quote password string

v0.8.6
------
- [COOK-534] - allow server_aliases to overridden by an attribute
- [COOK-799] - fixed disables .htaccess breaking permalink feature
- [COOK-820] - guard node.save with check for chef-solo in our cookbooks

v0.8.4
------
- [COOK-406] - wp-config.php.erb has wrong CRLF encoding
- Dropping explicit support for Red Hat platforms due to issues in php and mysql cookbooks (COOK-603, COOK-672, COOK-816, COOK-679)

v0.8.2
------
- [COOK-435] Don't set the mysql root user password in wordpress cookbook
- [COOK-535] - recursively create the directory
- RHEL/CentOS/Fedora support (yeah!)
- cleaned up node attribute keys
- cleaned up README.md


attributes/default.rb
#
# Author:: Barry Steinglass (<barry@opscode.com>)
# Author:: Koseki Kengo (<koseki@gmail.com>)
# Author:: Lucas Hansen (<lucash@opscode.com>)
# Author:: Julian C. Dunn (<jdunn@getchef.com>)
#
# Cookbook Name:: wordpress
# Attributes:: wordpress
#
# Copyright 2009-2013, Chef Software, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

# Helper defaults need to be initialized first
default['wordpress']['public_ip_query'] = 'wget http://ipecho.net/plain -O - -q'
extend Wordpress::Helpers

# General settings
default['wordpress']['version'] = 'latest'

default['wordpress']['db']['instance_name'] = 'default'
default['wordpress']['db']['name'] = "wordpressdb"
default['wordpress']['db']['user'] = "wordpressuser"
default['wordpress']['db']['prefix'] = 'wp_'
default['wordpress']['db']['hosts'] = %w{ localhost }
default['wordpress']['db']['port'] = '3306'  # Must be a string
default['wordpress']['db']['charset'] = 'utf8'
default['wordpress']['db']['collate'] = ''
case node['platform']
when 'ubuntu'
  case node['platform_version']
  when '10.04'
    default['wordpress']['db']['mysql_version'] = '5.1'
  else
    default['wordpress']['db']['mysql_version'] = '5.5'
  end
when 'centos', 'redhat', 'amazon', 'scientific'
  if node['platform_version'].to_i < 6
    default['wordpress']['db']['mysql_version'] = '5.0'
  elsif node['platform_version'].to_i < 7
    default['wordpress']['db']['mysql_version'] = '5.1'
  else
    default['wordpress']['db']['mysql_version'] = '5.7'
  end
else
  default['wordpress']['db']['mysql_version'] = '5.7'
end

default['wordpress']['allow_multisite'] = false

default['wordpress']['ssl_enabled'] = false

default['wordpress']['wp_config_options'] = {}

default['wordpress']['config_perms'] = 0644
default['wordpress']['server_aliases'] = [node['fqdn']]
default['wordpress']['server_port'] = '80'

default['wordpress']['install']['user'] = node['apache']['user']
default['wordpress']['install']['group'] = node['apache']['group']

# Languages
default['wordpress']['languages']['lang'] = ''
default['wordpress']['languages']['version'] = ''
default['wordpress']['languages']['repourl'] = 'http://translate.wordpress.org/projects/wp'
default['wordpress']['languages']['projects'] = ['main', 'admin', 'admin_network', 'continents_cities']
default['wordpress']['languages']['themes'] = []
default['wordpress']['languages']['project_pathes'] = {
  'main'              => '/',
  'admin'             => '/admin/',
  'admin_network'     => '/admin/network/',
  'continents_cities' => '/cc/'
}
%w{ten eleven twelve thirteen fourteen fifteen sixteen seventeen eighteen nineteen twenty}.each do |year|
  default['wordpress']['languages']['project_pathes']["twenty#{year}"] = "/twenty#{year}/"
end
node['wordpress']['languages']['project_pathes'].each do |project,project_path|
  # http://translate.wordpress.org/projects/wp/3.5.x/admin/network/ja/default/export-translations?format=mo
  default['wordpress']['languages']['urls'][project] =
    node['wordpress']['languages']['repourl'] + '/' +
    node['wordpress']['languages']['version'] + project_path +
    node['wordpress']['languages']['lang'] + '/default/export-translations?format=mo'
end

if node['platform'] == 'windows'
  default['wordpress']['parent_dir'] = "#{ENV['SystemDrive']}\\inetpub"
  default['wordpress']['dir'] = "#{node['wordpress']['parent_dir']}\\wordpress"
  default['wordpress']['url'] = "https://wordpress.org/wordpress-#{node['wordpress']['version']}.zip"
else
  default['wordpress']['server_name'] = node['fqdn']
  default['wordpress']['parent_dir'] = '/var/www'
  default['wordpress']['dir'] = "#{node['wordpress']['parent_dir']}/wordpress"
  default['wordpress']['url'] = "https://wordpress.org/wordpress-#{node['wordpress']['version']}.tar.gz"
end

default['wordpress']['php_options'] = { 'php_admin_value[upload_max_filesize]' => '50M', 'php_admin_value[post_max_size]' => '55M' }

# PHP
default['php']['fpm_user']      = node['apache']['user']
default['php']['fpm_group']     = node['apache']['group']
default['php']['fpm_listen_user'] = node['apache']['user']
default['php']['fpm_listen_group'] = node['apache']['group']

# wp-cli
default['wordpress']['wp_cli_url'] = 'https://raw.githubusercontent.com/wp-cli/builds/gh-pages/phar/wp-cli.phar'
default['wordpress']['wp_cli_bin'] = '/usr/local/bin/wp'

# Wordpress site
default['wordpress']['site']['url'] = "http://#{default_ip}"
default['wordpress']['site']['title'] = 'My WordPress Blog'
default['wordpress']['site']['admin_user'] = 'admin'
default['wordpress']['site']['admin_password'] = 'admin'
default['wordpress']['site']['admin_email'] = 'admin@example.com'

# Plugins
default['wordpress']['plugins'] = {
  ## Example by name: [id]: [plugin_name]
# disable_check_comment_flood: 'disable-check-comment-flood',
  ## Example by url: [id]: [plugin_url]
# fakerpress: 'https://github.com/bordoni/fakerpress/archive/0.3.1.zip'
}
